cachi2/core/package_managers/yarn_classic/resolver.py

```diff
@@ -48,26 +59,98 @@ class _RelpathMixin(BaseModel):
 class RegistryPackage(_BasePackage, _UrlMixin):
     """A Yarn 1.x package from the registry."""
 
+    @property
+    def purl(self) -> str:
+        """Return package URL."""
+        qualifiers = {}
+
+        if YARN_REGISTRY_URL in self.url:
+            qualifiers = {"repository_url": YARN_REGISTRY_URL}
+
+        if self.integrity:
+            qualifiers["checksum"] = str(ChecksumInfo.from_sri(self.integrity))
+
+        return PackageURL(
+            type="npm",
+            name=self.name,
+            version=self.version,
+            qualifiers=qualifiers,
+        ).to_string()
+
 
 class GitPackage(_BasePackage, _UrlMixin):
     """A Yarn 1.x package from a git repo."""
 
+    @property
+    def purl(self) -> str:
+        """Return package URL."""
+        parsed_url = urlparse(self.url)
+        repo_id = RepoID(origin_url=self.url, commit_id=parsed_url.fragment)
+        qualifiers = {"vcs_url": repo_id.as_vcs_url_qualifier()}
+        return PackageURL(
+            type="npm",
+            name=self.name,
+            version=self.version,
+            qualifiers=qualifiers,
+        ).to_string()
+
 
 class UrlPackage(_BasePackage, _UrlMixin):
     """A Yarn 1.x package from a http/https URL."""
 
+    @property
+    def purl(self) -> str:
+        """Return package URL."""
+        qualifiers = {"download_url": self.url}
+        return PackageURL(
+            type="npm",
+            name=self.name,
+            version=self.version,
+            qualifiers=qualifiers,
+        ).to_string()
+
 
 class FilePackage(_BasePackage, _RelpathMixin):
     """A Yarn 1.x package from a local file path."""
 
+    @property
+    def purl(self) -> str:
```

Comment:
Shouldn't this have a  `vcs_url` qualifier, since it lives within the source repo? Same comment for Link and Workspace packages

Reply:
I think so, we use `vcs_url` for PATH dependencies for [bundler](https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/package_managers/bundler/parser.py#L184).

Reply:
should be fixed